File: packages/template-android/app/src/main/java/com/rayact/engine/RayactHost.kt
package com.rayact.engine

import android.os.Handler
import android.os.Looper
import com.rayact.app.NavigationHost
import com.rayact.app.RayactRenderScheduler
import com.rayact.app.RayactScreenFragment
import com.rayact.app.RayactSurfaceView
import com.rayact.devclient.DevClientBridge
import com.rayact.devclient.DevMenuOverlay
import java.lang.ref.WeakReference
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class RayactHost(
    private val session: RayactEngineSession,
    val renderScheduler: RayactRenderScheduler
) : RayactEngineHostCallbacks {

    @Volatile
    private var navigationHost: WeakReference<NavigationHost>? = null

    @Volatile
    private var devMenuOverlay: DevMenuOverlay? = null

    @Volatile
    var imeView: RayactSurfaceView? = null
        private set

    val hostCallbacks: RayactEngineHostCallbacks get() = this

    fun registerImeView(v: RayactSurfaceView) { imeView = v }
    fun unregisterImeView(v: RayactSurfaceView) { if (imeView === v) imeView = null }

    fun setNavigationHost(h: NavigationHost) {
        navigationHost = WeakReference(h)
    }

    fun clearNavigationHost(h: NavigationHost) {
        if (navigationHost?.get() === h) navigationHost = null
    }

    override fun requestNewSurface(): Int {
        val h = navigationHost?.get() ?: return 0
        val latch = CountDownLatch(1)
        val result = IntArray(1)
        Handler(Looper.getMainLooper()).post {
            try {
                val frag = h.pushScreen()
                frag.onSurfaceReady = { sid ->
                    h.noteSurfaceReady(frag, sid)
                    result[0] = sid
                    latch.countDown()
                }
                if (frag.surfaceId > 0) {
                    result[0] = frag.surfaceId
                    latch.countDown()
                }
            } catch (_: Throwable) {
                latch.countDown()
            }
        }
        latch.await(2, TimeUnit.SECONDS)
        return result[0]
    }

    override fun releaseTopSurface() {
        val h = navigationHost?.get() ?: return
        Handler(Looper.getMainLooper()).post { h.popScreen() }
    }

    override fun releaseSurface(surfaceId: Int) {
        val h = navigationHost?.get() ?: return
        Handler(Looper.getMainLooper()).post { h.releaseSurface(surfaceId) }
    }

    override fun orderSurfaces(surfaceIds: IntArray) {
        val h = navigationHost?.get() ?: return
        Handler(Looper.getMainLooper()).post { h.orderSurfaces(surfaceIds) }
    }

    override fun rootSurfaceId(): Int = navigationHost?.get()?.rootSurfaceId ?: 0

    override fun topSurfaceId(): Int {
        val h = navigationHost?.get() ?: return 0
        return h.topFragmentSurfaceId().takeIf { it > 0 } ?: h.rootSurfaceId
    }

    override fun finishActivity() {
        // Return to the in-app dev launcher instead of finishing the Activity
        // whenever a project pane is actually open over a launcher session. This
        // is gated purely at runtime (tryShowLauncherFromFinishActivity checks a
        // dev host is registered and the active session isn't the launcher) — a
        // plain app is always in the launcher pane, so it still finishes. The old
        // BuildConfig.RAYACT_DEV_CLIENT compile gate disabled this in release
        // builds, so the release dev-app exited (and crashed in the two-session
        // onDestroy teardown) instead of going back to the launcher.
        if (DevClientBridge.tryShowLauncherFromFinishActivity()) {
            return
        }
        val h = navigationHost?.get() ?: return
        val activity = (h.context as? android.app.Activity) ?: return
        Handler(Looper.getMainLooper()).post { activity.finish() }
    }

    fun attachDevMenuOverlay(h: NavigationHost) {
        Handler(Looper.getMainLooper()).post {
            devMenuOverlay = DevMenuOverlay(h, session)
        }
    }

    override fun toggleDevMenu() {
        Handler(Looper.getMainLooper()).post { devMenuOverlay?.toggle() }
    }

    override fun requestRenderFrame() {
        renderScheduler.requestFrame()
    }

    override fun stopRenderScheduler() {
        renderScheduler.stopRendering()
    }

    override fun showSoftKeyboard(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Boolean,
        secure: Boolean,
        imeAction: String
    ) {
        val view = imeView ?: return
        Handler(Looper.getMainLooper()).post {
            view.setupForIme(nodeId, value, inputType, autocorrect, secure, imeAction)
        }
    }

    override fun switchIme(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Boolean,
        secure: Boolean,
        imeAction: String
    ) {
        val view = imeView ?: return
        Handler(Looper.getMainLooper()).post {
            view.switchIme(nodeId, value, inputType, autocorrect, secure, imeAction)
        }
    }

    override fun hideSoftKeyboard() {
        val view = imeView ?: return
        Handler(Looper.getMainLooper()).post { view.clearIme() }
    }

    override fun updateImeState(
        nodeId: Int,
        selectionStart: Int,
        selectionEnd: Int,
        composingStart: Int,
        composingEnd: Int,
        text: String?
    ) {
        val view = imeView ?: return
        Handler(Looper.getMainLooper()).post {
            view.updateImeState(nodeId, text, selectionStart, selectionEnd, composingStart, composingEnd)
        }
    }

    override fun copyToClipboard(text: String) {
        val view = imeView ?: return
        Handler(Looper.getMainLooper()).post {
            val clipboard = view.context.getSystemService(android.content.Context.CLIPBOARD_SERVICE)
                as? android.content.ClipboardManager
            clipboard?.setPrimaryClip(android.content.ClipData.newPlainText("rayact", text))
        }
    }

    override fun readClipboard(): String {
        val view = imeView ?: return ""
        val clipboard = view.context.getSystemService(android.content.Context.CLIPBOARD_SERVICE)
            as? android.content.ClipboardManager
        val clip = clipboard?.primaryClip ?: return ""
        if (clip.itemCount <= 0) return ""
        return clip.getItemAt(0).coerceToText(view.context)?.toString().orEmpty()
    }

    override fun performHapticFeedback() {
        val view = imeView ?: return
        Handler(Looper.getMainLooper()).post {
            view.performHapticFeedback(android.view.HapticFeedbackConstants.TEXT_HANDLE_MOVE)
        }
    }
}

File: packages/template-ios/RayactHost.swift
import UIKit

final class RayactHost: RayactEngineHostCallbacks {
    private let session: RayactEngineSession
    let renderScheduler: RayactRenderScheduler

    private weak var navigationHost: NavigationHost?
#if !RAYACT_RELEASE
    private var devMenuOverlay: DevMenuOverlay?
#endif
    private(set) weak var imeView: RayactSurfaceView?
    private var surfaceViews: [Int: WeakSurfaceView] = [:]
    private var pendingPlatformViewOperations: [Int: [PlatformViewOperation]] = [:]

    private final class WeakSurfaceView {
        weak var value: RayactSurfaceView?
        init(_ value: RayactSurfaceView) { self.value = value }
    }

    private enum PlatformViewOperation {
        case create(nodeId: Int, kind: String, propertiesJson: String)
        case properties(nodeId: Int, propertiesJson: String)
        case dispose(nodeId: Int)
    }

    init(session: RayactEngineSession, renderScheduler: RayactRenderScheduler) {
        self.session = session
        self.renderScheduler = renderScheduler
    }

    func registerImeView(_ view: RayactSurfaceView) { imeView = view }
    func unregisterImeView(_ view: RayactSurfaceView) {
        if imeView === view { imeView = nil }
    }

    func registerSurfaceView(_ view: RayactSurfaceView, surfaceId: Int) {
        surfaceViews[surfaceId] = WeakSurfaceView(view)
        let operations = pendingPlatformViewOperations.removeValue(
            forKey: surfaceId) ?? []
        for operation in operations {
            applyPlatformViewOperation(operation, to: view.platformViewHost)
        }
    }

    func unregisterSurfaceView(_ view: RayactSurfaceView, surfaceId: Int) {
        if surfaceViews[surfaceId]?.value === view {
            surfaceViews[surfaceId] = nil
            pendingPlatformViewOperations[surfaceId] = nil
        }
    }

    func setNavigationHost(_ host: NavigationHost) {
        navigationHost = host
    }

    func clearNavigationHost(_ host: NavigationHost) {
        if navigationHost === host { navigationHost = nil }
    }

    func requestNewSurface() -> Int {
        guard let nav = navigationHost else { return 0 }
        var result = 0
        let sem = DispatchSemaphore(value: 0)
        DispatchQueue.main.async {
            do {
                let screen = nav.pushScreen()
                screen.onSurfaceReady = { sid in
                    nav.noteSurfaceReady(screen, surfaceId: sid)
                    result = sid
                    sem.signal()
                }
                if screen.surfaceId > 0 {
                    result = screen.surfaceId
                    sem.signal()
                }
            } catch {
                sem.signal()
            }
        }
        _ = sem.wait(timeout: .now() + 2)
        return result
    }

    func releaseTopSurface() {
        guard let nav = navigationHost else { return }
        DispatchQueue.main.async { nav.popScreen() }
    }

    func releaseSurface(surfaceId: Int) {
        guard let nav = navigationHost else { return }
        DispatchQueue.main.async { nav.releaseSurface(surfaceId: surfaceId) }
    }

    func orderSurfaces(surfaceIds: [Int]) {
        guard let nav = navigationHost else { return }
        DispatchQueue.main.async { nav.orderSurfaces(surfaceIds: surfaceIds) }
    }

    func rootSurfaceId() -> Int {
        navigationHost?.rootSurfaceId ?? 0
    }

    func topSurfaceId() -> Int {
        guard let nav = navigationHost else { return 0 }
        let top = nav.topFragmentSurfaceId()
        return top > 0 ? top : nav.rootSurfaceId
    }

    func finishActivity() {
#if !RAYACT_RELEASE
        if DevClientBridge.tryShowLauncherFromFinishActivity() { return }
#endif
        guard let nav = navigationHost else { return }
        DispatchQueue.main.async {
            nav.parentController?.dismiss(animated: true)
        }
    }

    func attachDevMenuOverlay(_ host: NavigationHost) {
#if !RAYACT_RELEASE
        DispatchQueue.main.async {
            self.devMenuOverlay = DevMenuOverlay(host: host, session: self.session)
        }
#else
        _ = host
#endif
    }

    func toggleDevMenu() {
#if !RAYACT_RELEASE
        DispatchQueue.main.async { self.devMenuOverlay?.toggle() }
#endif
    }

    func requestRenderFrame() {
        renderScheduler.requestFrame()
    }

    func sendDevtoolsMessage(_ message: String) {
        session.sendDevtoolsMessage(message)
    }

    func stopRenderScheduler() {
        renderScheduler.stopRendering()
    }

    func showSoftKeyboard(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Bool,
        secure: Bool,
        imeAction: String,
        autoCapitalize: String,
        contextMenuHidden: Bool
    ) {
        guard let view = imeView else { return }
        DispatchQueue.main.async {
            view.setupForIme(
                nodeId: nodeId,
                initialText: value,
                inputType: inputType,
                autocorrect: autocorrect,
                secure: secure,
                imeAction: imeAction,
                autoCapitalize: autoCapitalize,
                contextMenuHidden: contextMenuHidden
            )
        }
    }

    func switchIme(
        nodeId: Int,
        value: String,
        inputType: String,
        autocorrect: Bool,
        secure: Bool,
        imeAction: String,
        autoCapitalize: String,
        contextMenuHidden: Bool
    ) {
        guard let view = imeView else { return }
        DispatchQueue.main.async {
            view.switchIme(
                nodeId: nodeId,
                initialText: value,
                inputType: inputType,
                autocorrect: autocorrect,
                secure: secure,
                imeAction: imeAction,
                autoCapitalize: autoCapitalize,
                contextMenuHidden: contextMenuHidden
            )
        }
    }

    func hideSoftKeyboard() {
        guard let view = imeView else { return }
        DispatchQueue.main.async { view.clearIme() }
    }

    func updateImeState(
        nodeId: Int,
        selectionStart: Int,
        selectionEnd: Int,
        composingStart: Int,
        composingEnd: Int,
        text: String?
    ) {
        guard let view = imeView else { return }
        DispatchQueue.main.async {
            view.updateImeState(
                nodeId: nodeId,
                text: text,
                selectionStart: selectionStart,
                selectionEnd: selectionEnd,
                composingStart: composingStart,
                composingEnd: composingEnd
            )
        }
    }

    func copyToClipboard(text: String) {
        guard let view = imeView else { return }
        DispatchQueue.main.async {
            UIPasteboard.general.string = text
        }
    }

    func readClipboard() -> String {
        UIPasteboard.general.string ?? ""
    }

    func performHapticFeedback() {
        guard let view = imeView else { return }
        DispatchQueue.main.async {
            view.performHapticFeedback()
        }
    }

    func platformViewCreate(surfaceId: Int, nodeId: Int, kind: String, propertiesJson: String) {
        DispatchQueue.main.async { [weak self] in
            self?.routePlatformViewOperation(
                .create(nodeId: nodeId, kind: kind,
                        propertiesJson: propertiesJson),
                surfaceId: surfaceId)
        }
    }

    func platformViewSetProperties(surfaceId: Int, nodeId: Int, propertiesJson: String) {
        DispatchQueue.main.async { [weak self] in
            self?.routePlatformViewOperation(
                .properties(nodeId: nodeId, propertiesJson: propertiesJson),
                surfaceId: surfaceId)
        }
    }

    func platformViewDispose(surfaceId: Int, nodeId: Int) {
        DispatchQueue.main.async { [weak self] in
            self?.routePlatformViewOperation(
                .dispose(nodeId: nodeId), surfaceId: surfaceId)
        }
    }

    private func routePlatformViewOperation(
        _ operation: PlatformViewOperation, surfaceId: Int
    ) {
        if let host = surfaceViews[surfaceId]?.value?.platformViewHost {
            applyPlatformViewOperation(operation, to: host)
        } else {
            pendingPlatformViewOperations[surfaceId, default: []].append(operation)
        }
    }

    private func applyPlatformViewOperation(
        _ operation: PlatformViewOperation, to host: RayactPlatformViewHost
    ) {
        switch operation {
        case let .create(nodeId, kind, propertiesJson):
            host.create(
                nodeId: nodeId, kind: kind, propertiesJson: propertiesJson)
        case let .properties(nodeId, propertiesJson):
            host.setProperties(
                nodeId: nodeId, propertiesJson: propertiesJson)
        case let .dispose(nodeId):
            host.dispose(nodeId: nodeId)
        }
    }

    func platformViewsBeginFrame(surfaceId: Int, width: Float, height: Float, density: Float) {
        surfaceViews[surfaceId]?.value?.platformViewHost.beginFrame(
            width: width, height: height, density: density)
    }

    func platformViewComposite(surfaceId: Int, nodeId: Int, compositionJson: String) -> Bool {
        surfaceViews[surfaceId]?.value?.platformViewHost.composite(
            nodeId: nodeId, compositionJson: compositionJson) ?? false
    }

    func platformViewsEndFrame(surfaceId: Int) {
        surfaceViews[surfaceId]?.value?.platformViewHost.endFrame()
    }

    func platformViewGestureDecision(surfaceId: Int, nodeId: Int, accepted: Bool) {
        surfaceViews[surfaceId]?.value?.platformViewHost.resolveGesture(
            nodeId: nodeId, accepted: accepted)
    }
}
